// react-chord-diagram/es/ChordDiagram.js
var _class, _temp;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { arc } from 'd3-shape';
import { ribbon, chord } from 'd3-chord';
import { scaleOrdinal } from 'd3-scale';
import { range, descending } from 'd3-array';

import Svg from './Svg';
import Groups from './Groups';
import Ribbons from './Ribbons';

import './main.css';

var ChordDiagram = (_temp = _class = function (_Component) {
    _inherits(ChordDiagram, _Component);

    function ChordDiagram(props) {
        _classCallCheck(this, ChordDiagram);

        var _this = _possibleConstructorReturn(this, _Component.call(this, props));

        _this.state = {
            hoverPersist: false,
            mouseOverGroup: null,
            mouseOverRibbon: null
        };


        _this.clearHover = _this.clearHover.bind(_this);
        _this.setHoverPersist = _this.setHoverPersist.bind(_this);
        _this.setMouseOverGroup = _this.setMouseOverGroup.bind(_this);
        _this.setMouseOverRibbon = _this.setMouseOverRibbon.bind(_this);
        return _this;
    }

    ChordDiagram.prototype.clearHover = function clearHover() {
        this.setState({ hoverPersist: false, mouseOverGroup: null, mouseOverRibbon: null });
    };

    ChordDiagram.prototype.setHoverPersist = function setHoverPersist(hoverPersist) {
        if (this.props.persistHoverOnClick) {
            this.setState({ hoverPersist: hoverPersist });
        }
    };

    ChordDiagram.prototype.setMouseOverGroup = function setMouseOverGroup(mouseOverGroup) {
        this.setState({ mouseOverGroup: mouseOverGroup });
    };

    ChordDiagram.prototype.setMouseOverRibbon = function setMouseOverRibbon(mouseOverRibbon) {
        this.setState({ mouseOverRibbon: mouseOverRibbon });
    };

    ChordDiagram.prototype.render = function render() {
        var _props = this.props,
            matrix = _props.matrix,
            componentId = _props.componentId,
            width = _props.width,
            height = _props.height,
            style = _props.style,
            className = _props.className,
            groupLabels = _props.groupLabels,
            groupColors = _props.groupColors,
            groupOnClick = _props.groupOnClick,
            padAngle = _props.padAngle,
            sortGroups = _props.sortGroups,
            sortSubgroups = _props.sortSubgroups,
            sortChords = _props.sortChords,
            labelColors = _props.labelColors,
            disableHover = _props.disableHover,
            disableGroupHover = _props.disableGroupHover,
            disableRibbonHover = _props.disableRibbonHover,
            strokeWidth = _props.strokeWidth,
            resizeWithWindow = _props.resizeWithWindow,
            ribbonOnClick = _props.ribbonOnClick,
            blurOnHover = _props.blurOnHover,
            ribbonOpacity = _props.ribbonOpacity,
            ribbonBlurOpacity = _props.ribbonBlurOpacity,
            persistHoverOnClick = _props.persistHoverOnClick,
            svgOnClick = _props.svgOnClick;


        var outerRadius = this.props.outerRadius || Math.min(width, height) * 0.5 - 40;
        var innerRadius = this.props.innerRadius || outerRadius - 30;

        var d3Chord = chord().padAngle(padAngle).sortGroups(sortGroups).sortSubgroups(sortSubgroups).sortChords(sortChords);

        var chords = d3Chord(matrix);

        var d3Arc = arc().innerRadius(innerRadius).outerRadius(outerRadius);

        var d3Ribbon = ribbon().radius(innerRadius);

        var color = scaleOrdinal().domain(range(groupColors.length)).range(groupColors);

        return React.createElement(
            Svg,
            {
                width: width,
                height: height,
                style: style,
                className: className,
                clearHover: this.clearHover,
                resizeWithWindow: resizeWithWindow,
                onClick: svgOnClick
            },
            React.createElement(Groups, {
                componentId: componentId,
                chords: chords,
                color: color,
                arc: d3Arc,
                outerRadius: outerRadius,
                setMouseOverGroup: this.setMouseOverGroup,
                groupLabels: groupLabels,
                labelColors: labelColors,
                disableHover: disableHover || disableGroupHover,
                hoverPersist: this.state.hoverPersist,
                setHoverPersist: this.setHoverPersist,
                onClick: groupOnClick
            }),
            React.createElement(Ribbons, {
                chords: chords,
                color: color,
                disableHover: disableHover || disableRibbonHover,
                ribbon: d3Ribbon,
                setMouseOverRibbon: this.setMouseOverRibbon,
                mouseOverGroup: this.state.mouseOverGroup,
                mouseOverRibbon: this.state.mouseOverRibbon,
                strokeWidth: strokeWidth,
                hoverPersist: this.state.hoverPersist,
                setHoverPersist: this.setHoverPersist,
                onClick: ribbonOnClick,
                blurOnHover: blurOnHover,
                ribbonOpacity: ribbonOpacity,
                ribbonBlurOpacity: ribbonBlurOpacity
            })
        );
    };

    return ChordDiagram;
}(Component), _class.defaultProps = {
    matrix: [],
    componentId: 1,
    width: 700,
    height: 700,
    style: {},
    className: '',
    outerRadius: null,
    innerRadius: null,
    groupLabels: [],
    groupColors: [],
    groupOnClick: null,
    padAngle: 0.05,
    sortGroups: null,
    sortSubgroups: descending,
    sortChords: null,
    labelColors: ['#000000'],
    disableHover: false,
    disableGroupHover: false,
    disableRibbonHover: true,
    strokeWidth: 1,
    resizeWithWindow: false,
    ribbonOnClick: null,
    blurOnHover: false,
    ribbonOpacity: '0.67',
    ribbonHoverOpacity: '0.2',
    persistHoverOnClick: false,
    svgOnClick: null
}, _temp);
export { ChordDiagram as default };
ChordDiagram.propTypes = process.env.NODE_ENV !== "production" ? {
    matrix: PropTypes.array.isRequired,
    componentId: PropTypes.number.isRequired,
    width: PropTypes.number,
    height: PropTypes.number,
    style: PropTypes.object,
    className: PropTypes.string,
    outerRadius: PropTypes.number,
    innerRadius: PropTypes.number,
    groupLabels: PropTypes.array,
    groupColors: PropTypes.array,
    padAngle: PropTypes.number,
    sortGroups: PropTypes.func,
    sortSubgroups: PropTypes.func,
    sortChords: PropTypes.func,
    labelColors: PropTypes.array,
    disableHover: PropTypes.bool,
    disableGroupHover: PropTypes.bool,
    disableRibbonHover: PropTypes.bool,
    strokeWidth: PropTypes.number,
    resizeWithWindow: PropTypes.bool,
    groupOnClick: PropTypes.func,
    ribbonOnClick: PropTypes.func,
    svgOnClick: PropTypes.func,
    blurOnHover: PropTypes.bool,
    ribbonOpacity: PropTypes.string,
    ribbonHoverOpacity: PropTypes.string,
    persistHoverOnClick: PropTypes.bool
} : {};
